### PR TITLE
gRPCLB and RR race fixes

### PR DESCRIPTION
**Pull request by dgquintas**
Misplaced locks and a big bug related to reusing closure instances: each `glb_pick` receives a callback. We wrap that callback, mainly to do reference-counting on the underlying RR instance. Previously, a single stack-allocated instance of `grpc_closure` was used. This is obviously (well, maybe not so obviously after all) wrong the moment we have concurrent picks.

### CODE REVIEW

File: src/core/ext/lb_policy/grpclb/grpclb.c [hunk 1]
```diff
@@ -158,6 +158,12 @@ typedef struct wrapped_rr_closure_arg {
   /* when not NULL, represents a pending_{pick,ping} node to be freed upon
    * closure execution */
   void *owning_pending_node; /* to be freed if not NULL */
+
+  /* Pointer ot heap memory if the closure and its argument were allocated
```
Comment (markdroth) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: s/ot/to/

File: src/core/ext/lb_policy/grpclb/grpclb.c [hunk 2]
```diff
@@ -186,7 +192,17 @@ static void wrapped_rr_closure(grpc_exec_ctx *exec_ctx, void *arg,
 
   grpc_exec_ctx_sched(exec_ctx, wc_arg->wrapped_closure, GRPC_ERROR_REF(error),
                       NULL);
-  gpr_free(wc_arg->owning_pending_node);
+
+  /* Make sure this closure and its arg are EITHER on the heap on their oen OR
```
Comment (markdroth) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: s/oen/own/

File: src/core/ext/lb_policy/grpclb/grpclb.c [hunk 3]
```diff
@@ -775,37 +788,50 @@ static int glb_pick(grpc_exec_ctx *exec_ctx, grpc_lb_policy *pol,
...
-      initial_metadata_add_lb_token(
-          pick_args->initial_metadata, pick_args->lb_token_mdelem_storage,
-          GRPC_MDELEM_REF(glb_policy->wc_arg.lb_token));
+      initial_metadata_add_lb_token(pick_args->initial_metadata,
+                                    pick_args->lb_token_mdelem_storage,
+                                    GRPC_MDELEM_REF(wc_arg->lb_token));
+
+      gpr_free(closure_mem);
     }
+    /* else, !pick_done, the pending pick will be registered and taken care of
```
Comment (markdroth) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: Please move this comment down below the "else", since it documents what happens in that block.

File: src/core/ext/lb_policy/grpclb/grpclb.c [hunk 4]
```diff
@@ -186,7 +192,17 @@ static void wrapped_rr_closure(grpc_exec_ctx *exec_ctx, void *arg,
 
   grpc_exec_ctx_sched(exec_ctx, wc_arg->wrapped_closure, GRPC_ERROR_REF(error),
                       NULL);
-  gpr_free(wc_arg->owning_pending_node);
+
+  /* Make sure this closure and its arg are EITHER on the heap on their oen OR
+   * part of a pending pick (thus part of the pending pick's memory) */
+  GPR_ASSERT((wc_arg->closure_mem_or_null != NULL) +
```
Comment (markdroth) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: Since we're treating both of these fields exactly the same (i.e., they both get freed), why bother having two different fields to begin with?  Why not just have a single `free_when_done` field, which gets set to either the closure or the pending pick, depending on where it's called from?

This may be moot if you take my suggestion above about storing the closure in wrapped_rr_closure_arg.

File: src/core/ext/lb_policy/grpclb/grpclb.c [hunk 5]
```diff
@@ -775,37 +788,50 @@ static int glb_pick(grpc_exec_ctx *exec_ctx, grpc_lb_policy *pol,
...
-    glb_policy->wc_arg.target = target;
-    glb_policy->wc_arg.wrapped_closure = on_complete;
-    glb_policy->wc_arg.lb_token_mdelem_storage =
-        pick_args->lb_token_mdelem_storage;
-    glb_policy->wc_arg.initial_metadata = pick_args->initial_metadata;
-    glb_policy->wc_arg.owning_pending_node = NULL;
-    grpc_closure_init(&glb_policy->wrapped_on_complete, wrapped_rr_closure,
-                      &glb_policy->wc_arg);
+
+    /* we need to allocate the closure on the stack because we may be serving
```
Comment (markdroth) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: s/stack/heap/

File: src/core/ext/lb_policy/grpclb/grpclb.c [hunk 6]
```diff
@@ -158,6 +158,12 @@ typedef struct wrapped_rr_closure_arg {
   /* when not NULL, represents a pending_{pick,ping} node to be freed upon
    * closure execution */
   void *owning_pending_node; /* to be freed if not NULL */
+
+  /* Pointer ot heap memory if the closure and its argument were allocated
+   * dynamically outside of a pending pick. It'll be NULL otherwise.
+   *
+   * TODO(dgq): This is by no means pretty. */
+  void *closure_mem_or_null;
```
Comment (markdroth) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: Here's an alternative suggestion that might be cleaner: Store the closure in wrapped_rr_closure_arg.  That way, we never have to free it at all, because it will be freed when wrapped_rr_closure_arg is.  This also avoids the duplication in the existing code where we currently store a closure in each of pending_pick and pending_ping.

File: src/core/ext/lb_policy/round_robin/round_robin.c [hunk 7]
```diff
@@ -416,6 +415,7 @@ static int rr_pick(grpc_exec_ctx *exec_ctx, grpc_lb_policy *pol,
     }
     /* only advance the last picked pointer if the selection was used */
     advance_last_picked_locked(p);
+    gpr_mu_unlock(&p->mu);
```
Comment (markdroth) [inline src/core/ext/lb_policy/round_robin/round_robin.c]: Why is this change needed?  I'm probably missing something obvious, but I'd like to make sure I understand the reason for the change.

File: src/core/ext/lb_policy/grpclb/grpclb.c (hunk 4 above)
Comment (dgquintas) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: Good idea, done.

We still need to do a free, even after taking your suggestion.

File: src/core/ext/lb_policy/grpclb/grpclb.c (hunk 6 above)
Comment (dgquintas) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: yes! Done.

File: src/core/ext/lb_policy/grpclb/grpclb.c [hunk 8]
```diff
@@ -474,29 +460,29 @@ static void rr_handover(grpc_exec_ctx *exec_ctx, glb_lb_policy *glb_policy,
     glb_policy->pending_picks = pp->next;
     GRPC_LB_POLICY_REF(glb_policy->rr_policy, "rr_handover_pending_pick");
     pp->wrapped_on_complete_arg.rr_policy = glb_policy->rr_policy;
+    pp->wrapped_on_complete_arg.free_when_done = pp;
```
Comment (markdroth) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: Instead of setting this here, how about doing it in `add_pending_pick()` above?  Even though we don't call it until we get to this point, we might as well initialize it properly from the get-go.

File: src/core/ext/lb_policy/grpclb/grpclb.c [hunk 9]
```diff
@@ -474,29 +460,29 @@ static void rr_handover(grpc_exec_ctx *exec_ctx, glb_lb_policy *glb_policy,
...
-    pp->wrapped_on_complete_arg.owning_pending_node = pp;
+                        &pp->wrapped_on_complete_arg.wrapper_closure);
   }
 
   pending_ping *pping;
   while ((pping = glb_policy->pending_pings)) {
     glb_policy->pending_pings = pping->next;
     GRPC_LB_POLICY_REF(glb_policy->rr_policy, "rr_handover_pending_ping");
     pping->wrapped_notify_arg.rr_policy = glb_policy->rr_policy;
+    pping->wrapped_notify_arg.free_when_done = pping;
```
Comment (markdroth) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: Same comment here -- this could be set in `add_pending_ping()`.

File: src/core/ext/lb_policy/grpclb/grpclb.c (hunk 8 above)
Comment (dgquintas) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: heh, I was making this change in parallel before reading your comment, same thoughts. Done.

File: src/core/ext/lb_policy/grpclb/grpclb.c (hunk 9 above)
Comment (dgquintas) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: Ditto.

File: src/core/ext/lb_policy/grpclb/grpclb.c [hunk 10]
```diff
@@ -155,9 +158,8 @@ typedef struct wrapped_rr_closure_arg {
   /* The RR instance related to the closure */
   grpc_lb_policy *rr_policy;
```
Comment (markdroth) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: I just noticed that this field is probably not needed.  We're setting it in a bunch of places, but it doesn't look like we're actually using it anywhere.  (Well, we are using it in a couple of places where we should be using `glb_policy->rr_policy` instead, but that's easily fixed.)

Comment (dgquintas) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: We use it in the closure function, on line 171 and following. The closure's arg doesn't have `glb_policy` pointer. Are you suggesting adding a `glb_policy` pointer to the closure arg in order to have access to `rr_policy`?

Comment (dgquintas) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: The reasoning behind having the `rr_policy` pointer instead of a `glb_policy` pointer in the closure arg was to limit how much was exposed to the arg.

Comment (markdroth) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: You're right -- I somehow missed that we were unreffing the RR policy there.  Sorry for the confusion.

File: src/core/ext/lb_policy/grpclb/grpclb.c [hunk 11]
```diff
@@ -484,8 +473,7 @@ static void rr_handover(grpc_exec_ctx *exec_ctx, glb_lb_policy *glb_policy,
     grpc_lb_policy_pick(exec_ctx, glb_policy->rr_policy, &pp->pick_args,
```
Comment (markdroth) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: What happens if this returns true (i.e., if the pick completes immediately)?  In that case, won't we fail to invoke (or clean up) the closure?

If it's somehow not possible for this to return true, then we should probably have an assert and a comment explaining why.

Comment (dgquintas) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: You are right. I guess it's unlikely that the RR policy, very recently created, would be ready to return anything immediately, which is why this hasn't been flushed out by tests.

Given that this is the second place where we pick from RR and have to branch on the result, I've factored out the logic into `pick_from_internal_rr()`.

Comment (markdroth) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: LGTM
